src/librustc_typeck/check/demand.rs

```diff
@@ -306,12 +305,248 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                                                  format!("*{}", code)));
                                 }
                             }
-                        },
+                        }
                     }
                 }
                 None
             }
             _ => None,
         }
     }
+
+    fn check_for_cast(&self,
+                      err: &mut DiagnosticBuilder<'tcx>,
+                      expr: &hir::Expr,
+                      checked_ty: Ty<'tcx>,
+                      expected_ty: Ty<'tcx>)
+                      -> bool {
+        let will_truncate = "will truncate the source value";
+        let depending_on_isize = "will truncate or zero-extend depending on the bit width of \
+                                  `isize`";
+        let depending_on_usize = "will truncate or zero-extend depending on the bit width of \
+                                  `usize`";
+        let will_sign_extend = "will sign-extend the source value";
+        let will_zero_extend = "will zero-extend the source value";
+
+        // If casting this expression to a given numeric type would be appropriate in case of a type
+        // mismatch.
+        //
+        // We want to minimize the amount of casting operations that are suggested, as it can be a
+        // lossy operation with potentially bad side effects, so we only suggest when encountering
+        // an expression that indicates that the original type couldn't be directly changed.
+        //
+        // For now, don't suggest casting with `as`.
+        let can_cast = false;
+
+        let needs_paren = expr_precedence(expr) < (AssocOp::As.precedence() as i8);
```

Comment:
Argh, this `expr_precedence` works on AST.
There's another one in `rustc::hir::print`.
Also, `AssocOp::As.precedence()` -> `PREC_POSTFIX`.

Reply:
I'm gonna add a commit to merge both into one place (new enum that only holds the type to precedence mapping, and an into impl for `{hir, ast}::Expr` so that there's only one source of truth for this. Started that yesterday but went to bed before I could complete it.